### PR TITLE
Security and Privacy

### PR DESCRIPTION
**Pull request by aarani**
The goal for this PR is to implement as much security checks we can.

### CODE REVIEW

**Comment by aarani** [comment]
Since this is a rabbit hole we need to flush these every few commits
so we don't make it impossible to review them.

File: NOnion/Directory/KeyCertificatesDocument.fs [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+                if decryptedDocumentDigest <> computedDocumentDigest then
+                    failwith
+                        "KeyCertificate validation failed: Certification validation failed"
+
+                // Clear history before we move on to next item
+                clearHistory() |> ignore<StringBuilder>
+
+                innerParse KeyCertificateEntry.Empty (state :: previousEntries)
+
+            if lines.Count = 0 then
```
Comment (knocte) [inline NOnion/Directory/KeyCertificatesDocument.fs:119]: let's use Any()

File: NOnion/Directory/NetworkStatusDocument.fs [hunk 2]
```diff
@@ -252,12 +257,25 @@ type DirectorySignature =
                 | "directory-signature" when state.Identity = None ->
                     lines.Dequeue() |> ignore<string>
 
-                    innerParse
-                        { state with
-                            DirectorySignature.Identity = readWord() |> Some
-                            Digest = readWord() |> Some
-                            Signature = readBlock String.Empty |> Some
-                        }
+                    let algs = [ "sha1"; "sha256" ]
```
Comment (knocte) [inline NOnion/Directory/NetworkStatusDocument.fs:260]: nit: rename to algos

File: NOnion/Directory/TorDirectory.fs [hunk 3]
```diff
@@ -43,6 +45,128 @@ type TorDirectory =
...
+                    match keyCertOpt with
+                    | Some keyCert ->
+                        let signingKey =
+                            UnwrapOption
+                                keyCert.SigningKey
+                                "ValidateConsensus: key cert's signing key is none (should not happen)"
+                            |> PemUtility.GetRsaKeyParametersFromPem
+
+                        let digest =
+                            match alg with
```
Comment (knocte) [inline NOnion/Directory/TorDirectory.fs:135]: @aarani nit: rename to algo

File: NOnion/Utility/FSharpUtil.fs [hunk 4]
```diff
@@ -95,3 +95,8 @@ module FSharpUtil =
             }
 
         retryLoop 0
+
+    let UnwrapOption<'T> (opt: Option<'T>) (msg: string) : 'T =
+        match opt with
+        | Some value -> value
+        | None -> failwith <| sprintf "error unwrapping Option: %s" msg
```
[the remark this answers is not in the document]
Comment (aarani) [reply NOnion/Utility/FSharpUtil.fs:102]: I don't agree, NOnion does not currently depend on fsxc and I don't like adding an entire library just for one function.
If in the future we needed more from fsxc, I'm open to adding it.

[the remark this answers is not in the document]
Comment (aarani) [reply NOnion/Utility/FSharpUtil.fs:102]: also as far as I can see fsxc is a .NET 6 package. We use .NET Standard so we can't consume it.

[the remark this answers is not in the document]
Comment (knocte) [reply NOnion/Utility/FSharpUtil.fs:102]: 2 things:

- I didn't mean for just this function sorry, I mean remove the whole FSharpUtil.fs and use Fsxc.
- We would depend on Fsdk, not fsx (my bad, mistyped).

File: NOnion/Directory/KeyCertificatesDocument.fs (hunk 1 above)
Comment (aarani) [reply NOnion/Directory/KeyCertificatesDocument.fs]: I used Seq.isEmpty instead.

File: NOnion/Directory/NetworkStatusDocument.fs (hunk 2 above)
Comment (aarani) [reply NOnion/Directory/NetworkStatusDocument.fs]: Fixed

File: NOnion/Directory/TorDirectory.fs (hunk 3 above)
Comment (aarani) [reply NOnion/Directory/TorDirectory.fs:135]: Fixed

File: NOnion/Network/TorGuard.fs [hunk 5]
```diff
@@ -383,7 +383,7 @@ type TorGuard private (client: TcpClient, sslStream: SslStream) =
                             CellNetInfo.Time =
                                 DateTimeUtils.ToUnixTimestamp DateTime.UtcNow
                             OtherAddress = otherAddress
-                            MyAddresses = List.singleton netInfo.OtherAddress
+                            MyAddresses = List.empty
```
Comment (knocte) [inline NOnion/Network/TorGuard.fs]: @aarani nit: use List.Empty, not List.empty

Comment (aarani) [reply NOnion/Network/TorGuard.fs]: Fixed

**Comment by knocte** [comment]
typo in one of the commit msgs: validatation

**Comment by knocte** [comment]
> Directory: fix broken directory-signature parsing

Why was it broken?

File: NOnion.Tests/DirectoryParsers.cs [hunk 6]
```diff
@@ -49,5 +49,19 @@ public void CanParseMicroDescriptorAndConvertToJson()
 
             Assert.That(microDescriptorJson, Is.EqualTo(expectedMicroDescriptorsJson));
         }
+
+        [Test]
+        public void CanParseKeyCertificateAndConvertToJson()
+        {
+            // Tor directory spec enforces documents to use \n (The ascii LF character (hex value 0x0a)
```
Comment (knocte) [inline NOnion.Tests/DirectoryParsers.cs:56]: forgot to close parentheses

File: NOnion/Constants.fs [hunk 7]
```diff
@@ -197,3 +197,7 @@ module Constants =
 
     let internal CertificateCertifiedKeyLength = 32
     let internal CertificateSignatureLength = 64
+
+    module CertTypes =
+        let Link = 1uy
+        let ID = 2uy
```
Comment (knocte) [inline NOnion/Constants.fs:203]: shouldn't this be a DU=Enum?

Comment (knocte) [reply NOnion/Constants.fs:203]: ah no, nevermind

File: NOnion/Directory/KeyCertificatesDocument.fs [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+                        { state with
+                            Certification = readBlock String.Empty |> Some
+                        }
+                        historyBeforeCertification
+                | _ ->
+                    //Ignore possible unknown items (future-proofness)
+                    dequeueLineAndAddToHistory() |> ignore<string>
+                    innerParse state previousEntries
+
+        innerParse KeyCertificateEntry.Empty List.empty
```
Comment (knocte) [inline NOnion/Directory/KeyCertificatesDocument.fs:215]: @aarani should be List.Empty

File: NOnion/Utility/EmbeddedResourceUtility.fs [hunk 9]
```diff
@@ -0,0 +1,40 @@
+﻿namespace NOnion.Utility
+
+open System
+open System.IO
+open System.Reflection
+
+module EmbeddedResourceUtility =
+
+    // Code from https://github.com/nblockchain/geewallet/blob/428cb77d21dba20fc38c7ea032003c5861aac950/src/GWallet.Backend/Config.fs#L156
+    let ExtractEmbeddedResourceFileContents(resourceName: string) : string =
```
Comment (knocte) [inline NOnion/Utility/EmbeddedResourceUtility.fs:10]: @aarani I wonder if we should put this in Fsdk too

File: NOnion.Tests/DirectoryParsers.cs (hunk 6 above)
Comment (aarani) [reply NOnion.Tests/DirectoryParsers.cs]: Fixed

File: NOnion/Utility/EmbeddedResourceUtility.fs (hunk 9 above)
Comment (aarani) [reply NOnion/Utility/EmbeddedResourceUtility.fs:10]: > > Directory: fix broken directory-signature parsing
> 
> Why was it broken?

Added to the commit msg

**Comment by aarani** [comment]
> > Directory: fix broken directory-signature parsing
> 
> Why was it broken?

Added to the commit msg

File: NOnion/Utility/EmbeddedResourceUtility.fs (hunk 9 above)
Comment (aarani) [reply NOnion/Utility/EmbeddedResourceUtility.fs:10]: Good idea though

**Review by knocte — APPROVED** [review]

**Pull request merged**